stdlib/atexit.pyi

```diff
@@ -1,7 +1,10 @@
 from typing import Any, Callable
+from typing_extensions import ParamSpec
+
+_P = ParamSpec("_P")
 
 def _clear() -> None: ...
 def _ncallbacks() -> int: ...
 def _run_exitfuncs() -> None: ...
-def register(func: Callable[..., Any], *args: Any, **kwargs: Any) -> Callable[..., Any]: ...
-def unregister(func: Callable[..., Any]) -> None: ...
+def register(func: Callable[_P, Any], *args: Any, **kwargs: Any) -> Callable[_P, Any]: ...  # type: ignore
```

Comment:
It might be better to use a TypeVar, since it returns the function as is. This way, you can say "it returns the exact same type as the first argument" instead of "it returns a callable that takes the same arguments".

```suggestion
def register(func: _T, *args: Any, **kwargs: Any) -> _T: ...
```

This would prevent getting errors from code like this, for example:

```python
class MyCallable:
    def foo(self):
        print("Foo")
    def __call__(self):
        pass

atexit.register(MyCallable()).foo()
```

To ensure  that non-callables are not accepted, you can define `_T` as `TypeVar("_T", bound=Callable[..., Any])`.

Reply:
Actually, ignore that :D I think it would be more useful to say that `*args` and `**kwargs` must match the callable:

```suggestion
def register(func: Callable[_P, Any], *args: _P.args, **kwargs: _P.kwargs) -> Callable[_P, Any]: ...  # type: ignore
```

Reply:
We can probably use a TypeVar for the return value to get rid of the Any, right?

Reply:
I thought I understood, how `ParamSpec` works, but need to dive in a bit deeper and maybe I find some other good place to implement it 💪

Reply:
@JelleZijlstra yeah you are right

Reply:
@JelleZijlstra Any thoughts about whether `TypeVar` or `ParamSpec` is more suitable for this?  On the one hand, a `TypeVar` provides more accurate information about the return type, but on the other hand, a `ParamSpec` would ensure that you pass in compatible `*args` and `**kwargs`.

Reply:
Even I'm not so experienced with that topic, `ParamSpec` seems a better fit here with the `ParamSpecArgs` and `...Kwargs` notation.

Reply:
so at the end, it would look like this
```python
_T = TypeVar("_T")
_P = ParamSpec("_P")

def register(func: Callable[_P, _T], *args: _P.args, **kwargs: _P.kwargs) -> Callable[_P, _T]: ...  # type: ignore
```

Reply:
I put my suggestion in a separate comment because this thread is based on outdated code. We can both get a precise return type and use the ParamSpec to get precise types for the args and kwargs.